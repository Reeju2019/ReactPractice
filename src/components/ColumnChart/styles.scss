.strong {
  font-weight: bold;
}

.header-1-cchart {
  font-size: 1.8rem;
  font-weight: 700;
  line-height: 1.2;
}

.cchart-container-border {
  border: #dcdcdc solid;
  width: 50%;
}

.cchart-border-bottom {
  border-bottom: #dcdcdc 2px solid;
  margin: 0;
}
