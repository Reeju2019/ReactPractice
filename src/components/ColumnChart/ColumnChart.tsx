import React from "react";
import {
  Chart as ChartJS,
  CategoryScale,
  LinearScale,
  BarElement,
  Title,
  Tooltip,
  Legend,
} from "chart.js";
import { Col, Container, Form, Row } from "react-bootstrap";
import { Bar } from "react-chartjs-2";

import "./styles.scss";

/**
 * @description Props of Column Chart Component
 * @memberof ColumnChart
 * @property {any} localeConfig Sets the locale config data
 * @property {any} columnChartConfig Sets the column chart config data
 * @property {string} header Sets the header text
 * @property {any} layout Set the layout
 * @property {boolean} showLegend Set weather to show legend or not
 */
/* eslint-disable */
interface IProps {
  localeConfig: any;
  columnChartConfig: any;
  header: string;
  layout: any;
  showLegend?: boolean;
}

/**
 * @description State of Column Chart Component
 * @memberof ColumnChart
 * @property {any} columnChartConfig Set the column chart config data
 * @property {any} selectedModel Set the selected model
 */
/* eslint-disable */
interface IState {
  columnChartConfig: any;
  selectedModel: any;
}

ChartJS.register(CategoryScale, LinearScale, BarElement);

/**
 * @component
 * Thos component was used in App component
 * @description ColumnChart component
 */
class ColumnChart extends React.Component<IProps, IState> {
  state: IState = {
    selectedModel: null,
    columnChartConfig: null,
  };

  /**
   * @constructor
   * @description Constructor
   * @param {IProps} props Props
   */
  constructor(props: any) {
    super(props);
    this.bindThisToMethods = this.bindThisToMethods.bind(this);
    this.bindThisToMethods();
  }

  /**
   * @description Bind all the methods to this
   * @property {function} bindThisToMethods Bind This To Methods
   * @returns void
   */
  bindThisToMethods() {
    this.setConfig = this.setConfig.bind(this);
    this.selectModelHandler = this.selectModelHandler.bind(this);
  }

  /**
   * @description Set state according to the json configuration
   * @property {Function} setConfig Set Config
   * @returns void
   */
  setConfig() {
    const columnChart = this.props.columnChartConfig;
    columnChart.options["plugins"] = {
      legend: {
        display: !!this.props.showLegend,
      },
    };
    this.setState({
      columnChartConfig: columnChart,
      selectedModel: columnChart.data[0],
    });
  }

  /**
   * @description The state will be set to the json configuration here
   * @property {Function} componentDidMount Component Did Mount
   */
  componentDidMount(): void {
    this.setConfig();
  }

  /**
   * @description OnClick on select model handler to change UI to selected model
   * @property {Function} selectModelHandler Select Model Handler
   * @param {object} event Event
   * @returns void
   */
  selectModelHandler(event) {
    console.log(event.target.value);

    const model = this.state.columnChartConfig.data.filter(
      (item) => item.modelName === event.target.value
    );
    this.setState({ selectedModel: model[0] });
  }

  render() {
    return (
      <>
        {this.state.columnChartConfig && (
          <Row className="cchart-container-border m-5 m-5">
            <Row className="cchart-border-bottom py-4">
              <Col
                xs={this.props.layout.header.xs}
                sm={this.props.layout.header.sm}
                md={this.props.layout.header.md}
                lg={this.props.layout.header.lg}
                className="header-1-cchart"
              >
                {this.props.header}
              </Col>
              <Col
                xs={this.props.layout.dropdown.xs}
                sm={this.props.layout.dropdown.sm}
                md={this.props.layout.dropdown.md}
                lg={this.props.layout.dropdown.lg}
              >
                <Form.Select
                  aria-label="Default select example"
                  multiple={this.state.columnChartConfig.multiple}
                  onChange={(event) => this.selectModelHandler(event)}
                >
                  {this.state.columnChartConfig.data.map(
                    (item: any, index: number) => (
                      <option key={index}>{item.modelName}</option>
                    )
                  )}
                </Form.Select>
              </Col>
            </Row>
            <Row
              xs={this.props.layout.chart.xs}
              sm={this.props.layout.chart.sm}
              md={this.props.layout.chart.md}
              lg={this.props.layout.chart.lg}
              className="p-4 mt-2"
            >
              <Bar
                options={this.state.columnChartConfig.options}
                data={this.state.selectedModel}
              />
            </Row>
          </Row>
        )}
      </>
    );
  }
}

export default ColumnChart;
export { ColumnChart };
