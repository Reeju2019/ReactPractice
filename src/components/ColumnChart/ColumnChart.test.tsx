import { shallow } from "enzyme"
import ReactDOM from "react-dom"
import ColumnChart from "./ColumnChart"

const localeConfig = {
  uploadAsset_header_1: "Upload your",
  uploadAsset_header_2: "Asset",
  uploadAsset_subHeader_1: "Supported File Formats:",
  uploadAsset_subHeader_2: "img, jpeg, png and etc",
  uploadAsset_dropBoxHeader_1: "Drop your files here or",
  uploadAsset_dropBoxHeader_2: "Browse",
  uploadAsset_dropBoxSubHeader_1: "Max size of Video: 25 MB",
  uploadAsset_upload_msg_success: "uploading Complete!",
  uploadAsset_upload_msg_failure:
    "something went wrong, please check file type/size and upload again!",
}

const data = {
  header: "Grade Reports",
  data: [
    {
      modelName: "model1",
      labels: ["A+", "A", "B", "C", "D", "E"],
      datasets: [
        {
          label: "",
          backgroundColor: "#c1b4e7",
          barThickness: 20,
          maxBarThickness: 25,
          minBarLength: 20,
          barPercentage: 1.0,
          categoryPercentage: 1.0,
          data: [10, 20, 30, 40, 50, 60],
        },
      ],
    },
    {
      modelName: "model2",
      labels: ["Aa+", "Aa", "Bb", "Cc", "Dd", "Ee"],
      datasets: [
        {
          label: "",
          backgroundColor: "#000",
          barThickness: 20,
          maxBarThickness: 25,
          minBarLength: 20,
          barPercentage: 1.0,
          categoryPercentage: 1.0,
          data: [10, 20, 30, 40, 50, 60],
        },
      ],
    },
    {
      modelName: "model3",
      labels: ["A*", "A+", "B*", "C--", "D", "E"],
      datasets: [
        {
          label: "",
          backgroundColor: "#c1b4e7",
          barThickness: 20,
          maxBarThickness: 25,
          minBarLength: 20,
          barPercentage: 1.0,
          categoryPercentage: 1.0,
          data: [10, 20, 30, 40, 50, 60],
        },
      ],
    },
  ],
  options: {
    responsive: false,
  },
  multiple: false,
}

const layout = {
  header: {
    xs: 9,
    sm: 8,
    md: 6,
    lg: 6,
  },
  dropdown: {
    xs: 3,
    sm: 4,
    md: 6,
    lg: 6,
  },
  chart: {
    xs: 12,
    sm: 12,
    md: 12,
    lg: 12,
  },
}

/**
 * Unit testing for Column Chart Component
 */
describe("Column Chart Testing", () => {
  it("render without crashing", () => {
    const div = document.createElement("div")
    ReactDOM.render(
      <ColumnChart
        localeConfig={localeConfig}
        columnChartConfig={data}
        header={"grade report"}
        layout={layout}
      />,
      div
    )
    ReactDOM.unmountComponentAtNode(div)
  })
  it("should render initial layout", () => {
    const component = shallow(
      <ColumnChart
        localeConfig={localeConfig}
        columnChartConfig={data}
        header={"grade report"}
        layout={layout}
      />
    )
    expect(component.getElements()).toMatchSnapshot()
  })
})
