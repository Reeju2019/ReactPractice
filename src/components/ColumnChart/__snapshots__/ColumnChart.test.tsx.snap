// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Column Chart Testing should render initial layout 1`] = `
Array [
  <React.Fragment>
    <Row
      className="cchart-container-border m-5 m-5"
    >
      <Row
        className="cchart-border-bottom py-4"
      >
        <Col
          className="header-1-cchart"
          lg={6}
          md={6}
          sm={8}
          xs={9}
        >
          grade report
        </Col>
        <Col
          lg={6}
          md={6}
          sm={4}
          xs={3}
        >
          <FormSelect
            aria-label="Default select example"
            multiple={false}
            onChange={[Function]}
          >
            <option>
              model1
            </option>
            <option>
              model2
            </option>
            <option>
              model3
            </option>
          </FormSelect>
        </Col>
      </Row>
      <Row
        className="p-4 mt-2"
        lg={12}
        md={12}
        sm={12}
        xs={12}
      >
        <ForwardRef
          data={
            Object {
              "datasets": Array [
                Object {
                  "backgroundColor": "#c1b4e7",
                  "barPercentage": 1,
                  "barThickness": 20,
                  "categoryPercentage": 1,
                  "data": Array [
                    10,
                    20,
                    30,
                    40,
                    50,
                    60,
                  ],
                  "label": "",
                  "maxBarThickness": 25,
                  "minBarLength": 20,
                },
              ],
              "labels": Array [
                "A+",
                "A",
                "B",
                "C",
                "D",
                "E",
              ],
              "modelName": "model1",
            }
          }
          options={
            Object {
              "plugins": Object {},
              "responsive": false,
              "scales": Object {
                "x": Object {
                  "axis": "x",
                  "beginAtZero": false,
                  "bounds": "ticks",
                  "display": true,
                  "grace": 0,
                  "grid": Object {
                    "borderColor": "rgba(0,0,0,0.1)",
                    "borderDash": Array [],
                    "borderDashOffset": 0,
                    "borderWidth": 1,
                    "color": "rgba(0,0,0,0.1)",
                    "display": true,
                    "drawBorder": true,
                    "drawOnChartArea": true,
                    "drawTicks": true,
                    "lineWidth": 1,
                    "offset": true,
                    "tickColor": [Function],
                    "tickLength": 8,
                    "tickWidth": [Function],
                  },
                  "offset": true,
                  "reverse": false,
                  "ticks": Object {
                    "align": "center",
                    "autoSkip": true,
                    "autoSkipPadding": 3,
                    "backdropColor": "rgba(255, 255, 255, 0.75)",
                    "backdropPadding": 2,
                    "callback": [Function],
                    "color": "#666",
                    "crossAlign": "near",
                    "display": true,
                    "labelOffset": 0,
                    "major": Object {},
                    "maxRotation": 50,
                    "minRotation": 0,
                    "minor": Object {},
                    "mirror": false,
                    "padding": 3,
                    "showLabelBackdrop": false,
                    "textStrokeColor": "",
                    "textStrokeWidth": 0,
                  },
                  "title": Object {
                    "color": "#666",
                    "display": false,
                    "padding": Object {
                      "bottom": 4,
                      "top": 4,
                    },
                    "text": "",
                  },
                  "type": "category",
                },
                "y": Object {
                  "axis": "y",
                  "beginAtZero": true,
                  "bounds": "ticks",
                  "display": true,
                  "grace": 0,
                  "grid": Object {
                    "borderColor": "rgba(0,0,0,0.1)",
                    "borderDash": Array [],
                    "borderDashOffset": 0,
                    "borderWidth": 1,
                    "color": "rgba(0,0,0,0.1)",
                    "display": true,
                    "drawBorder": true,
                    "drawOnChartArea": true,
                    "drawTicks": true,
                    "lineWidth": 1,
                    "offset": false,
                    "tickColor": [Function],
                    "tickLength": 8,
                    "tickWidth": [Function],
                  },
                  "offset": false,
                  "reverse": false,
                  "ticks": Object {
                    "align": "center",
                    "autoSkip": true,
                    "autoSkipPadding": 3,
                    "backdropColor": "rgba(255, 255, 255, 0.75)",
                    "backdropPadding": 2,
                    "callback": [Function],
                    "color": "#666",
                    "crossAlign": "near",
                    "display": true,
                    "labelOffset": 0,
                    "major": Object {},
                    "maxRotation": 50,
                    "minRotation": 0,
                    "minor": Object {},
                    "mirror": false,
                    "padding": 3,
                    "showLabelBackdrop": false,
                    "textStrokeColor": "",
                    "textStrokeWidth": 0,
                  },
                  "title": Object {
                    "color": "#666",
                    "display": false,
                    "padding": Object {
                      "bottom": 4,
                      "top": 4,
                    },
                    "text": "",
                  },
                  "type": "linear",
                },
              },
            }
          }
        />
      </Row>
    </Row>
  </React.Fragment>,
]
`;
